pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -268,52 +308,139 @@ func (p *DNSProxy) RemoveRestoredRules(endpointID uint16) {
 	p.removeRestoredRulesLocked(uint64(endpointID))
 }
 
-// setPortRulesForID sets the matching rules for endpointID and destPort for
-// later lookups. It converts newRules into a unified regexp that can be reused
-// later.
-func (allow perEPAllow) setPortRulesForID(endpointID uint64, destPort uint16, newRules policy.L7DataMap) error {
-	// This is the delete case
-	if len(newRules) == 0 {
-		epPorts := allow[endpointID]
-		delete(epPorts, destPort)
-		if len(epPorts) == 0 {
-			delete(allow, endpointID)
+func (allow perEPAllow) addDestPortRules(endpointID uint64, destPort uint16, newRules policy.L7DataMap) error {
+	newCachedSelectorEntries := make(CachedSelectorEntry, len(newRules))
+	var err error
+	for selector, newRuleset := range newRules {
+		regexPattern, fqdnsList := GenerateRegexpAndFqdns(newRuleset)
+		var ruleEntry RegexpAndFqdnsEntry
+		ruleEntry, err = allow.addSingleRule(regexPattern, fqdnsList)
+		if err != nil {
+			break
 		}
-		return nil
+		newCachedSelectorEntries[selector] = ruleEntry
 	}
-
-	newRE, err := GetSelectorRegexMap(newRules)
 	if err != nil {
+		// Unregister all the registered rulesets before exiting
+		for _, ruleset := range newCachedSelectorEntries {
+			allow.removeSingleRule(ruleset)
+		}
 		return err
-	}
 
-	epPorts, exist := allow[endpointID]
+	}
+	allow.removeDestPortRules(endpointID, destPort)
+	epPorts, exist := allow.allowMap[endpointID]
 	if !exist {
 		epPorts = make(portToSelectorAllow)
-		allow[endpointID] = epPorts
+		allow.allowMap[endpointID] = epPorts
 	}
-
-	epPorts[destPort] = newRE
+	epPorts[destPort] = newCachedSelectorEntries
 	return nil
 }
 
+func (allow perEPAllow) removeSingleRule(value RegexpAndFqdnsEntry) {
+	if regexpIndexEntry, ok := allow.regexpIndex[value.regexp.String()]; ok {
+		if regexpIndexEntry.counter == 1 {
+			delete(allow.regexpIndex, value.regexp.String())
+		} else {
+			regexpIndexEntry.counter -= 1
+		}
```

Comment:
We could potentially define methods on these structs as helpers to hide some of these details behind a function call, to make the surrounding code a bit easier to read.